### PR TITLE
Add rule: strict-string-expressions

### LINKED ISSUE #4512: Rule suggestion: strict-string-expressions (no implicit function to string)
### Rule Suggestion

There is some common pattern

```js
const barFunc = () => 'World';
const fooStr = `Hello, ${barFunc()}`; // Hello, World
```

But it's easy to miss brackets around function and this would be not an error

```js
const barFunc = () => 'World';
const fooStr = `Hello, ${barFunc}`; // "Hello, () => 'World'"
```

**What does your suggested rule do?**

If it's `on`, require explicit `toString` or `String` call around `function`

Like this

```js
// yeah, we really want function source code here
const barFunc = () => 'World';
const fooStr = `Hello, ${String(barFunc)}`; // "Hello, () => 'World'"
```

**Additional context**

What about spreading this rule for other data types?

I think this is frustrating sometimes:

```js
const name = { foo: 'Foo' };
alert('Hello, ' + name); // Hello, [object Object] (╯°□°）╯︵ ┻━┻ 
```

What do you think?

### PR DESCRIPTION
**Pull request by ColCh**
#### PR checklist

- [x] Addresses an existing issue: fixes #4512
- [x] New feature, bugfix, or enhancement
  - [x] Includes tests
- [x] Documentation update `???`

#### Overview of change:

adds implementation for rule discussed in #4512

#### CHANGELOG.md entry:

[new-rule] `strict-string-expressions` reports errors on type coercions found in string expressions

### CODE REVIEW

File: src/rules/strictStringExpressionsRule.ts [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+            The following nodes are checked:
+
+            * String literals ("foo" + bar)
+            * ES6 templates (\`foo \${bar}\`)`,
+        type: "functionality",
+        typescriptOnly: true,
+        requiresTypeInfo: true,
+        options: [],
+        optionExamples: [true],
+        optionsDescription: "Not configurable.",
```
Comment (ColCh) [inline src/rules/strictStringExpressionsRule.ts]: I think there is no options at all

File: src/rules/strictStringExpressionsRule.ts [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        hasFix: true
+    };
+
+    public static CONVERSION_REQUIRED = 'Explicit conversion to string type required';
```
Comment (ColCh) [inline src/rules/strictStringExpressionsRule.ts]: error message is always same

File: src/rules/strictStringExpressionsRule.ts [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+            * String literals ("foo" + bar)
+            * ES6 templates (\`foo \${bar}\`)`,
+        type: "functionality",
+        typescriptOnly: true,
+        requiresTypeInfo: true,
+        options: [],
+        optionExamples: [true],
+        optionsDescription: "Not configurable.",
+        hasFix: true
```
Comment (ColCh) [inline src/rules/strictStringExpressionsRule.ts]: fixer just wraps expression into `String(FOO)` call to make it explicit

File: src/rules/strictStringExpressionsRule.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    public applyWithProgram(sourceFile: ts.SourceFile, program: ts.Program): Lint.RuleFailure[] {
+        return this.applyWithFunction(sourceFile, walk, undefined, program.getTypeChecker());
+    }
+}
+
+function walk(ctx: Lint.WalkContext<undefined>, checker: ts.TypeChecker): void {
+    const { sourceFile } = ctx;
+    ts.forEachChild(sourceFile, function cb(node: ts.Node): void {
+        switch (node.kind) {
+            case ts.SyntaxKind.BinaryExpression: {
```
Comment (ColCh) [inline src/rules/strictStringExpressionsRule.ts]: may be, this one can be omitted? due to `restrictPlusOperandsRule`

**Comment by JoshuaKGoldberg** [comment]
@ColCh looks like you've got some lint failures from CI running this rule on TSLint itself. https://circleci.com/gh/palantir/tslint/16921?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

That looks like it's exposed an edge case or two. Should `| null` and `| undefined` be allowed in string expressions? I think so; it's common for error logging to have strings like:

```js
function complain(error?: Error | null) {
    console.error(`Oh dear: ${error}`);
}
```

`new Error().stack` being `string | undefined` is also pretty annoying.

If you really wanted, you could make an issue to add an option to disallow this, or file a followup issue after this PR... but IMO it's fine to just always allow these.

**Comment by ColCh** [comment]
Yes, I agree with you. That `undefined` in union as well as other empty-in-runtime values like `never` or `null` creates so much problems :D

I think that better option is make those as configurable thing

File: src/rules/strictStringExpressionsRule.ts (hunk 1 above)
Comment (ColCh) [reply src/rules/strictStringExpressionsRule.ts]: oops. at least one now

File: src/rules/strictStringExpressionsRule.ts (hunk 4 above)
Comment (ColCh) [reply src/rules/strictStringExpressionsRule.ts]: I think not, because they do different things about `plus` operator

File: src/linter.ts [hunk 5]
```diff
@@ -187,7 +187,7 @@ export class Linter {
             this.options.formatter !== undefined ? this.options.formatter : "prose";
         const Formatter = findFormatter(formatterName, this.options.formattersDirectory);
         if (Formatter === undefined) {
-            throw new Error(`formatter '${formatterName}' not found`);
+            throw new Error(`formatter '${String(formatterName)}' not found`);
```
Comment (ColCh) [inline src/linter.ts]: real case here 🎉
image

**Comment by ColCh** [comment]
@JoshuaKGoldberg seems to pass now

I don't understand why `ci/circleci: testNext ` is failing but it's failing in master branch too

Screenshot 2019-07-28 at 22 09 03

so... please review :)

File: src/rules/strictStringExpressionsRule.ts [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+            walk,
+            this.getRuleOptions(),
+            program.getTypeChecker(),
+        );
+    }
+
+    private getRuleOptions(): Options {
+        if (this.ruleArguments[0] === undefined) {
+            return {
+                [OPTION_ALLOW_EMPTY_TYPES]: true,
```
Comment (ColCh) [inline src/rules/strictStringExpressionsRule.ts]: empty types are enabled by default

Comment (ColCh) [reply src/rules/strictStringExpressionsRule.ts]: I'm not sure about name though.. can `never` be considered as `empty`?

Comment (JoshuaKGoldberg) [reply src/rules/strictStringExpressionsRule.ts]: Great semantic question. Technically `never` is a [https://en.wikipedia.org/wiki/Bottom_type](bottom type); however, `void`, `undefined`, etc. are separate types all on their own. `empty` seems like a good way to refer to all these.

**Comment by ColCh** [comment]
Thank you!

File: src/rules/strictStringExpressionsRule.ts (hunk 6 above)
Comment (ColCh) [reply src/rules/strictStringExpressionsRule.ts]: yeah, that was difficult to find common name for these

**Pull request merged**